ArduCopter/mode_circle.cpp

```diff
@@ -27,41 +27,30 @@ bool Copter::ModeCircle::init(bool ignore_checks)
 // should be called at 100hz or more
 void Copter::ModeCircle::run()
 {
-    float target_yaw_rate = 0;
-    float target_climb_rate = 0;
-
     // initialize speeds and accelerations
     pos_control->set_max_speed_xy(wp_nav->get_default_speed_xy());
     pos_control->set_max_accel_xy(wp_nav->get_wp_acceleration());
     pos_control->set_max_speed_z(-get_pilot_speed_dn(), g.pilot_speed_up);
     pos_control->set_max_accel_z(g.pilot_accel_z);
-    
-    // if not auto armed or motor interlock not enabled set throttle to zero and exit immediately
-    if (!motors->armed() || !ap.auto_armed || ap.land_complete || !motors->get_interlock()) {
-        // To-Do: add some initialisation of position controllers
-        zero_throttle_and_relax_ac();
-        pos_control->set_alt_target_to_current_alt();
-        return;
+
+    // get pilot's desired yaw rate (or zero if in radio failsafe)
+    float target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+    if (!is_zero(target_yaw_rate)) {
+        pilot_yaw_override = true;
+    }
+
+    // get pilot desired climb rate (or zero if in radio failsafe)
+    float target_climb_rate = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
+    // adjust climb rate using rangefinder
+    if (copter.rangefinder_alt_ok()) {
+        // if rangefinder is ok, use surface tracking
+        target_climb_rate = get_surface_tracking_climb_rate(target_climb_rate, pos_control->get_alt_target(), G_Dt);
     }
 
-    // process pilot inputs
-    if (!copter.failsafe.radio) {
-        // get pilot's desired yaw rate
-        target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
-        if (!is_zero(target_yaw_rate)) {
-            pilot_yaw_override = true;
-        }
-
-        // get pilot desired climb rate
-        target_climb_rate = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
-
-        // check for pilot requested take-off
-        if (ap.land_complete && target_climb_rate > 0) {
-            // indicate we are taking off
-            set_land_complete(false);
-            // clear i term when we're taking off
-            set_throttle_takeoff();
-        }
+    // if not armed set throttle to zero and exit immediately
+    if (!motors->armed() || !ap.auto_armed || ap.land_complete) {
+        make_safe_spool_down();
```

Comment:
why not copter.circle_nav->init(); ?

Reply:
Thanks again for the review - it's very much appreciated because this is quite a bit RP.

So we are discussing what should happen when the vehicle is landed in Circle mode.  This is a real edge case and could only happen if:

- the user lands the vehicle by pulling down on the throttle stick
- the vehicle was landed in some other mode (like Loiter) and the user then switches to Circle

Calling circle_nav->init() would cause the center of the circle to be reset depending upon the vehicle's speed (it's not moving so it would be directly ahead of the vehicle).  It's not totally obvious if this is what we should do - if the user has not taken off yet then I agree it's probably the correct thing to do.  If the user has already been flying in Circle and has landed and is then taking off again then they might not expect the circle center to be moved.

In either case though, we don't support taking off in Circle mode.  Even if the users pulls the throttle stick up, the vehicle will not take off so in the end, they will be forced to switch to another mode, takeoff and then switch back into circle mode which will cause circle_nav->init() to be called.